_min_max_val.dim: CPU implementation

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42957 quant: switch observers to use min_max
* #44001 rename _min_max to _aminmax
* #42943 _min_max.dim: CUDA implementation
* **#42894 _min_max_val.dim: CPU implementation**
* #42868 min_max kernel: add CUDA

Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Note: after a discussion with @ngimel, we are taking the fast path
of calculating the values only and not the indices, since that is what
is needed for quantization, and calculating indices would require support
for reductions on 4 outputs which is additional work.  So, the API
doesn't fully match `min.dim` and `max.dim`.

Flexible on the name, let me know if something else is better.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23086798](https://our.internmc.facebook.com/intern/diff/D23086798)

vkuzo committed:
_min_max.dim: CPU implementation

Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
nit on "_min_max.dim: CPU implementation"


Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
remove indices on "_min_max.dim: CPU implementation"


Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "_min_max_val.dim: CPU implementation"


Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Note: after a discussion with @ngimel, we are taking the fast path
of calculating the values only and not the indices, since that is what
is needed for quantization, and calculating indices would require support
for reductions on 4 outputs which is additional work.  So, the API
doesn't fully match `min.dim` and `max.dim`.

Flexible on the name, let me know if something else is better.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ngimel commented:
min/max functions that don't calculate indices are going to be exposed as amin/amax #43092, so maybe it's better to call this aminmax.

ngimel commented on aten/src/ATen/native/cpu/TensorCompareKernel.cpp:
```diff
@@ -143,6 +143,50 @@ static void max_kernel_impl(
+            }
+          } else if (!(value <= max_number)) {
+            max_number = value;
+            if (_isnan<scalar_t>(value)) {
```
you can't ever get to this branch, `nan` is handled above

vkuzo replied:
thanks, fixed!

ngimel commented on test/test_torch.py:
```diff
@@ -15495,8 +15495,34 @@ def test_min(self, device, dtype):
-        self._test_minmax_helper(lambda x: torch._min_max(x)[0], np.min, device, dtype, skip_indices=True)
-        self._test_minmax_helper(lambda x: torch._min_max(x)[1], np.max, device, dtype, skip_indices=True)
+
+        def _min_wrapper(x, dim=None, keepdims=False):
```
You never call this with `dim` that is not None and  keepdims=True?

vkuzo replied:
the dims are cycled through in `_test_minmax_helper`.

however, I didn't know about `np.amin/np.amax` when writing this code. After seeing the PR you linked to we can just rename `torch._min_max` to `torch._aminmax` to simplify this further.  If it's ok, let me do that in a separate PR.

ngimel replied:
One would think I should know that since I wrote _test_minmax_helper :-) Yeah, let's do renaming in a separate PR (and then another renaming to `aminmax` without underscore when everything is ready).

vkuzo replied:
fixed naming in https://github.com/pytorch/pytorch/pull/44001, further up the stack

ngimel reviewed (commented):
Please make sure that tests are actually testing added functionality.

vkuzo committed:
Update on "_min_max_val.dim: CPU implementation"


Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Note: after a discussion with @ngimel, we are taking the fast path
of calculating the values only and not the indices, since that is what
is needed for quantization, and calculating indices would require support
for reductions on 4 outputs which is additional work.  So, the API
doesn't fully match `min.dim` and `max.dim`.

Flexible on the name, let me know if something else is better.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23086798](https://our.internmc.facebook.com/intern/diff/D23086798)

[ghstack-poisoned]

vkuzo committed:
simplify test on "_min_max_val.dim: CPU implementation"


Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Note: after a discussion with @ngimel, we are taking the fast path
of calculating the values only and not the indices, since that is what
is needed for quantization, and calculating indices would require support
for reductions on 4 outputs which is additional work.  So, the API
doesn't fully match `min.dim` and `max.dim`.

Flexible on the name, let me know if something else is better.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23086798](https://our.internmc.facebook.com/intern/diff/D23086798)

[ghstack-poisoned]

ngimel reviewed (approved):
Looks good subject to CI, thank you!

vkuzo committed:
fix nit on "_min_max_val.dim: CPU implementation"


Summary:

Continuing the min_max kernel implementation, this PR adds the
CPU path when a dim is specified.  Next PR will replicate for CUDA.

Note: after a discussion with @ngimel, we are taking the fast path
of calculating the values only and not the indices, since that is what
is needed for quantization, and calculating indices would require support
for reductions on 4 outputs which is additional work.  So, the API
doesn't fully match `min.dim` and `max.dim`.

Flexible on the name, let me know if something else is better.

Test Plan:

correctness:
```
python test/test_torch.py TestTorchDeviceTypeCPU.test_minmax_cpu_float32
```

performance: seeing a 49% speedup on a min+max tensor with similar shapes
to what we care about for quantization observers (bench:
https://gist.github.com/vkuzo/b3f24d67060e916128a51777f9b89326). For
other shapes (more dims, different dim sizes, etc), I've noticed a
speedup as low as 20%, but we don't have a good use case to optimize
that so perhaps we can save that for a future PR.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23086798](https://our.internmc.facebook.com/intern/diff/D23086798)

[ghstack-poisoned]
